### DISCUSSION TITLE
Fit Panel Behavior Proposal

### DISCUSSION THREAD

**Discussion by butlerpd** [SasView/sasview#2190] [Ideas]
As mentioned in my comments to PR #1798, I think the behavior when calculating a theory or sending data to a Fit Panel is not optimal for a number of reasons and would like to start a discussion on that with this issue (which is listed in the "discussion" milestone at this point.
 
### Theory curve only in Fit Panel
1. Choosing a category and then a model brings up the model parameters with default parameters set but nothing else happens. Note: there is a disabled button labeled `Show Plot`. This label **changes** (bad form) to `Calculate` as it is enabled by the choice of an actual model with default parameters.
2. clicking the now active `Calculate` button does almost nothing at all from the point of view of the user other than discretely changing the label of the calculate button back to `Show Plot`.
   1. As a user, if I click a button I expect something useful to happen. In this case a plot.  If I really and truly wanted to change parameters **before** I a plot gets generated I would have changed them **before** hitting any button!
   2. Extra clicks in general should be minimized for good UI design
   3. Magically changing the label/function of a button on the UI is generally something one tries to avoid unless absolutely necessary. I would argue that is absolutely not the case here.
3. **PROPOSAL:** I would argue that both extra clicks are unnecessary and that in fact, as soon as I choose a model that has valid parameters (hopefully the defaults are valid) it should compute and plot.  I believe that even if this proposal is not accepted it should at the very least do the computation and be prepared to plot -- this last I would consider an honest bug.  The advantage of the proposal, aside from being what I, and I believe most actual users, would expect and want, is that one can then revisit PR #1798   and hopefully simplify the code since the button label would never have to change value and would always behave the same: recalculate with the current parameters and show/update the plot.

### Data sent to Fit Panel
1. In this case it seems that sending a data set to fitting will either
   1. If data is sent to a Fit Panel which already has a model chosen, it clearly forces a compute as well. However the button label remains `Calculate`. But pressing that button performs the actions of `Show Plot` and in fact the label does, at this point, also change to read `Show Plot`. So here the behavior of the `Calculate` button has magically changed (again - bad form).
   2. If data is sent to a Fit Panel with no model chosen, it will make it such that choosing a model with valid parameters (the defaults) will in fact force a compute. This behavior is now different from doing the same thing without a data set (not good practice I would argue).  In this case however, the disabled `Show Plot` button becomes enabled without changing value so that it at least is behaving the same way the button normally behaves.
2. **PROPOSAL:** Again, I would propose that both from the point of you of usability and from the point of view of code simplification the behavior should be that:
   1. As soon as a model is chosen it should computer and generate a plot
   2. As soon as a data set is sent to a Fit Panel it should be plotted, regardless of the existence of a model or  not. This would make it consistent with the `Create New` plot functionality. Further, from the point of view of a researcher, the entire point of using the GUI is to see my data as I work with it.
      
I believe this was implemented as a way to reduce the graph clutter that is a big complaint in SasView. However the problem is this seems to me the exact wrong thing to remove.  This in fact is the primary thing I want to see which is being obstructed by all the other plots (clutter), not the other way round.  So all the plots from **other** Fit Panels and or the distribution and residual plots are what are getting in the way of seeing what I'm working on in *this* fit panel. I always want to see what I am currently working on.  Unfortunately the other plots are also important so getting rid of them is not an option.  We need better plot management tools and have some tickets to start addressing some of those, such as the idea of putting the residuals on the same plot as the fit and/or adding an option to show/hide them as we do for error bars and putting plots like distributions in a different tabs on the fit plots etc.

In the meantime, I would like to make the case for this proposal which I see as a) always making the data in the panel I'm currently working on always be "the focus" and "on top" while b) simplifying the code and the user interface.

**Comment by smk78** [comment]
> 3. **PROPOSAL:** I would argue that both extra clicks are unnecessary and that in fact, as soon as I choose a model that has valid parameters (hopefully the defaults are valid) it should compute and plot.

Accepting this proposal would emulate the behaviour in 4.x, I believe.

**Comment by rozyczko** [comment]
**Theory curve**

> 2. clicking the now active `Calculate` button does almost nothing at all from the point of view of the user other than discretely changing the label of the calculate button back to Show Plot.

This action generates a new theory dataset, which is immediately visible in the data explorer.
By default this dataset is not being plotted, since, as you noticed, we originally decided on generating plots as a separate user-driven action, in order not to populate the workspace with "unwanted" plots.

> Extra clicks in general should be minimized for good UI design

Of course this is correct. However, it will depend on the expected workflow of operations a typical user performs. The decision was based mostly on generic UI suggestions, e.g. a click should perform one, well defined action only. 
If users expect/prefer to have the plots generated at the same time as the dataset is created then I would argue we need to replace the changing `Calculate` <-> `Show Plot` caption to something relevant, like `Generate Dataset` or `Add Dataset`.

**Data sent to Fit Panel**
> i. If data is sent to a Fit Panel which already has a model chosen, it clearly forces a compute as well. However the button label remains `Calculate`.

This seems to be buggy behaviour. Sending data to a page with model chosen should trigger recalculation and switching button caption to `Show Plot`.

> ii. (...) This behavior is now different from doing the same thing without a data set (not good practice I would argue). 

Not really. With the theory curve, you need to generate a new model instance and add it to the data explorer. This is the `Calculate` action. Each subsequent parameter change will force recalculation and update of the existing dataset model.

With actual data in the fit panel, choosing a model is performing the second part of the above - forcing recalculation and updating the **existing** model, hence we don't need "Calculate" anymore.


I think what you are presenting is a good way to make the GUI behave as it did in 4.x and this might be preferable to many of our users.
However, the current design is by no means random even if sometimes it might seem awkward or counter-intuitive.
The overall principle we chose was to make sure the user gets what they are being told they would, rather than perform several, often unrelated, operations with one user-action in order to save on clicks/button presses.
However, we also chose to mimic 4.x as much as possible and this constrained the design to some extent.
Maybe indeed making the GUI interaction match that of 4.x is better?
